File: shared/contracts/audit.ts
export interface AuditEvent {
  id: number;
  at: string;
  actorId: string | null;
  actorName: string;
  dspId: string | null;
  dspName: string | null;
  action: string;
  detail: string;
  area: AuditArea;
  target: string | null;
  // The record the event is about, when it has one that outlives a rename.
  ref: { kind: 'member' | 'role' | 'schedule' | 'job'; id: string } | null;
  changes: AuditChange[];
}
export type AuditArea =
  | 'team'
  | 'roles'
  | 'collections'
  | 'schedules'
  | 'connections'
  | 'access'
  | 'dsps'
  | 'settings';
// A changed field; a granted value has no `from` and a revoked one has no `to`.
export interface AuditChange {
  field: string;
  from: string | null;
  to: string | null;
}
export interface AuditPage {
  events: AuditEvent[];
  total: number;
  counts: Partial<Record<AuditArea | 'failures', number>>;
  actors: { id: string; name: string }[];
  // DSPs with activity, for narrowing the platform's log; empty inside a DSP.
  dsps: { id: string; name: string }[];
}

File: backend/src/contracts/audit.rs
use super::*;
use std::collections::BTreeMap;

text_enum! {
    #[cfg_attr(test, derive(ts_rs::TS))]
    pub enum AuditArea {
        Team => "team", Roles => "roles", Collections => "collections", Schedules => "schedules",
        Connections => "connections", Access => "access", Dsps => "dsps", Settings => "settings",
    }
}
text_enum! {
    #[cfg_attr(test, derive(ts_rs::TS))]
    pub enum AuditSubject { Member => "member", Role => "role", Schedule => "schedule", Job => "job", }
}
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub struct AuditReference {
    pub kind: AuditSubject,
    pub id: String,
}
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub struct AuditChange {
    pub field: String,
    pub from: Option<String>,
    pub to: Option<String>,
}
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
#[serde(rename_all = "camelCase")]
pub struct AuditEvent {
    #[cfg_attr(test, ts(type = "number"))]
    pub id: i64,
    pub at: String,
    pub actor_id: Option<String>,
    pub actor_name: String,
    pub dsp_id: Option<String>,
    pub dsp_name: Option<String>,
    pub action: String,
    pub detail: String,
    pub area: AuditArea,
    pub target: Option<String>,
    #[serde(rename = "ref")]
    pub reference: Option<AuditReference>,
    pub changes: Vec<AuditChange>,
}
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub struct AuditName {
    pub id: String,
    pub name: String,
}
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
pub struct AuditPage {
    pub events: Vec<AuditEvent>,
    #[cfg_attr(test, ts(type = "number"))]
    pub total: i64,
    pub counts: BTreeMap<AuditCount, usize>,
    pub actors: Vec<AuditName>,
    pub dsps: Vec<AuditName>,
}

text_enum! {
    #[cfg_attr(test, derive(ts_rs::TS))]
    #[derive(PartialOrd, Ord)]
    pub enum AuditCount {
        Team => "team", Roles => "roles", Collections => "collections", Schedules => "schedules",
        Connections => "connections", Access => "access", Dsps => "dsps", Settings => "settings", Failures => "failures",
    }
}
